node/core/candidate-validation/src/lib.rs

```diff
@@ -165,6 +168,63 @@ async fn run(
 						Err(e) => return Err(e),
 					}
 				}
+				CandidateValidationMessage::PreValidate(
+					para_id,
+					relay_parent,
+					params,
+					response_sender,
+				) => {
+					let (code_tx, code_rx) = oneshot::channel();
+					// TODO ensure that we needn't call this every time
+					let validation_code = match runtime_api_request(
```

Comment:
What is the best option to ensure the validation_code is only fetched if it has not yet been prepared?

Reply:
That's actually a feature in itself.

See https://github.com/paritytech/polkadot/issues/2866